src/inference/trace_translators.jl

```diff
@@ -0,0 +1,874 @@
+    first_pass_results = run_first_pass(f, prev_model_trace, forward_proposal_trace)
+    log_abs_determinant = jacobian_correction(
+        f, prev_model_trace, forward_proposal_trace, first_pass_results, nothing)
+    constraints = merge(first_pass_results.constraints, new_observations)
+    (new_model_trace, _) = generate(p_new, p_new_args, constraints)
+    backward_proposal_trace, = generate(
+        q_backward, (new_model_trace, q_backward_args...), first_pass_results.u_back)
+    return (new_model_trace, backward_proposal_trace, log_abs_determinant)
+end
+
+function (translator::GeneralTraceTranslator)(
+        prev_model_trace::Trace; check=false, prev_observations=EmptyChoiceMap())
+
+    # sample auxiliary trace
+    forward_proposal_trace = simulate(proposal, (prev_model_trace, translator.q_forward_args...,))
+
+    # apply trace transform
+    (new_model_trace, backward_proposal_trace, log_abs_determinant) = general_trace_translator_run_transform(
+        translator.f, prev_model_trace, forward_proposal_trace, translator.new_observations,
+        translator.p_new, translator.p_new_args, translator.q_backward, translator.q_backward_args)
+
+    # compute log weight
+    prev_model_score = get_score(prev_model_trace)
+    new_model_score = get_score(new_model_trace)
+    forward_proposal_score = get_score(forward_proposal_trace)
+    backward_proposal_score = get_score(backward_proposal_trace)
+    log_weight = new_model_score - prev_model_score + backward_proposal_score + forward_proposal_score + log_abs_determinant
+
+    if check
+        forward_proposal_choices = get_choices(forward_proposal_trace)
+        f_inv = inverse(translator.f)
+        (prev_model_trace_rt, forward_proposal_trace_rt, _) = general_trace_translator_run_transform(
+            inverse(translator.f), prev_observations,
+            new_model_trace, backward_proposal_trace,
+            get_gen_fn(prev_model_trace), get_args(prev_model_trace),
+            translator.q_forward, translator.q_forward_args)
+        check_round_trip(
+            prev_model_trace, prev_model_trace_rt,
+            forward_proposal_trace, forward_proposal_trace_rt)
+    end
+
+    return (new_model_trace, log_weight)
+end
+
+##################################
+# SimpleExtendingTraceTranslator #
+##################################
+
+"""
+    translator = SimpleExtendingTraceTranslator(;
+        p_new_args::Tuple = (),
+        argdiffs::Tuple = (),
+        new_obs::ChoiceMap = EmptyChoiceMap(),
+        q_fwd::GenerativeFunction,
+        q_fwd_args::Tuple  = ())
+
+Constructor for a simple extending trace translator.
+
+Run the translator with:
+
+    (output_trace, log_weight) = translator(input_trace)
+"""
+@with_kw struct SimpleExtendingTraceTranslator 
+    p_new_args::Tuple = ()
+    argdiffs::Tuple = ()
+    new_obs::ChoiceMap = EmptyChoiceMap()
+    q_fwd::GenerativeFunction
+    q_fwd_args::Tuple  = ()
+end
```

Comment:
I wonder if it might be possible to also support a "less simple" `ExtendingTraceTranslator", which allows users to provide a `TraceTransformDSLProgam" to map choices generated by `q_fwd` to choices within `p_new`? This might arise in cases where the proposal doesn't sample choices that also exist within the extended model. For example, in a time series model, where the models trace addresses are nested within a namespace like `1` within an `Unfold` combinator, it might be nice to write a proposal which doesn't also have to nest its choices within that namespace, and instead specify a trace transform to do the appropriate mapping.

Reply:
This will also be necessary if `q_fwd` is itself a `Map` or `Unfold` combinator, in which case a trace transform would be required to remap all the indices from `1, ..., t` to `T+1, ..., T+t`.

Reply:
Yes, that sounds useful. I think it's separable from this PR, so let's make it an issue.

Reply:
@ztangent Here is an issue to track this: https://github.com/probcomp/Gen.jl/issues/313